src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs

```diff
@@ -0,0 +1,301 @@
+                        {
+                            delimited.BeginDelimitedElement();
+                            tableExpression.NormalizedPredicate.AcceptVisitor(tableExpression.SearchParameterQueryGenerator, this);
+                        }
+                    }
+
+                    break;
+
+                case TableExpressionKind.Concatenation:
+                    StringBuilder.Append("SELECT Sid1 FROM ").AppendLine(TableExpressionName(_tableExpressionCounter - 1));
+                    StringBuilder.AppendLine("UNION ALL");
+
+                    goto case TableExpressionKind.Normal;
+
+                case TableExpressionKind.All:
+                    StringBuilder.Append("SELECT ").Append(V1.Resource.ResourceSurrogateId).AppendLine(" AS Sid1")
+                        .Append("FROM ").AppendLine(V1.Resource);
+
+                    using (var delimited = StringBuilder.BeginDelimitedWhereClause())
+                    {
+                        AppendHistoryClause(delimited);
+                        AppendDeletedClause(delimited);
+                        if (tableExpression.DenormalizedPredicate != null)
+                        {
+                            delimited.BeginDelimitedElement();
+                            tableExpression.DenormalizedPredicate?.AcceptVisitor(this, context);
+                        }
+                    }
+
+                    break;
+
+                case TableExpressionKind.NotExists:
+                    StringBuilder.Append("SELECT Sid1 FROM ").AppendLine(TableExpressionName(_tableExpressionCounter - 1));
+                    StringBuilder.AppendLine("WHERE Sid1 NOT IN").AppendLine("(");
+
+                    using (StringBuilder.Indent())
+                    {
+                        StringBuilder.Append("SELECT ").AppendLine(V1.Resource.ResourceSurrogateId)
+                            .Append("FROM ").AppendLine(tableExpression.SearchParameterQueryGenerator.Table);
+                        using (var delimited = StringBuilder.BeginDelimitedWhereClause())
+                        {
+                            AppendHistoryClause(delimited);
+
+                            if (tableExpression.DenormalizedPredicate != null)
+                            {
+                                delimited.BeginDelimitedElement();
+                                tableExpression.DenormalizedPredicate?.AcceptVisitor(this, context);
+                            }
+
+                            delimited.BeginDelimitedElement();
+                            tableExpression.NormalizedPredicate.AcceptVisitor(tableExpression.SearchParameterQueryGenerator, this);
+                        }
+                    }
+
+                    StringBuilder.AppendLine(")");
+                    break;
+
+                case TableExpressionKind.Top:
+                    StringBuilder.Append("SELECT DISTINCT TOP ").Append(context.MaxItemCount + 1).AppendLine(" Sid1 ")
```

Comment:
We should be able to use a param instead (e.g. top (@maxItemCount)) to increase reuse of cached plans.